[12.0][MIG] auth_saml

gurneyalex opened the pull request.

max3903 committed:
[IMP] Move files in a subdirectory

[IMP] Cleanup

max3903 committed:
[MIG] auth_saml: Migration to 11.0

yvaucher committed:
Typo

gurneyalex committed:
[MIG] auth_saml to 12.0

add requirement on lasso

pedrobaeza commented:
@probabble @alexey-pelykh you can now review this one.

gurneyalex commented:
caution: PR is hot from the oven, and not tested yet with a real SAML server. The code installs, the views display AFAICT. I'll be testing soon to assert if this works, but I need to get my hands on my customer's sysadmin :smile_cat: In the meanwhile if you are able to test this in real, I'll be more than happy

oca-travis committed:
[UPD] Update auth_saml.pot

blop commented:
Testing the auth_saml in v12 :)

blop commented:
Have anyone succeeded in making this module work on a v12 ?

OCA-git-bot committed:
[ADD] icon.png

pedrobaeza committed:
[FIX] auth_saml: Remove broken symlink

moylop260 committed:
[FIX] auth_saml: Fix KeyError using auth_oauth module. (#149)

The following line of code for 11.0:
 - https://github.com/odoo/odoo/blob/52d6f0e3ee90874fc93fec9cdff74ec71d3b991f/addons/auth_oauth/controllers/main.py#L69

is assigning the key "auth_link" for "list_providers" method.

The following template is expecting this key:
 - https://github.com/odoo/odoo/blob/52d6f0e3ee90874fc93fec9cdff74ec71d3b991f/addons/auth_oauth/views/auth_oauth_templates.xml#L5

So, it raise a KeyError compiling "template_auth_oauth_providers_N"

This change is fixing adding that expected key in order to avoid this KeyError

oca-travis committed:
[UPD] Update auth_saml.pot

oca-transbot committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: server-auth-11.0/server-auth-11.0-auth_saml
Translate-URL: https://translation.odoo-community.org/projects/server-auth-11-0/server-auth-11-0-auth_saml/

Commit:
[FIX] dependencies

[FIX] dependencies

yajo commented on auth_saml/LICENSE:
```diff
@@ -0,0 +1,661 @@
+                    GNU AFFERO GENERAL PUBLIC LICENSE
```
:wastebasket: No need to include LICENSE here AFAIK.

yajo commented on auth_saml/README.rst:
```diff
@@ -0,0 +1,119 @@
+
+====================
+SAML2 authentication
+====================
```
Please [split the README](https://github.com/OCA/maintainer-tools/tree/becfd07b7fdac38cb02a86a522c82d11b1c8e6fd/template/module/readme) and autogenerate it.

yajo commented on auth_saml/__manifest__.py:
```diff
@@ -0,0 +1,29 @@
+    'category': 'Tools',
+    'author': 'XCG Consulting, Odoo Community Association (OCA)',
+    'maintainer': 'XCG Consulting',
+    'website': 'http://odoo.consulting',
```
```suggestion
    'website': 'https://github.com/OCA/server-auth',
```

yajo commented on auth_saml/controllers/main.py:
```diff
@@ -0,0 +1,216 @@
+    def get_auth_request(self, pid):
+        """state is the JSONified state object and we need to pass
+        it inside our request as the RelayState argument
+        """
```
This docstring has nothing to do with the method content. Is it updated?

yajo commented on auth_saml/controllers/main.py:
```diff
@@ -0,0 +1,216 @@
+            auth_request = provider._get_auth_request(state)
+
+        except Exception as e:
+            _logger.exception("SAML2: %s" % str(e))
```
```suggestion
            _logger.exception("SAML2: %s", e)
```

yajo commented on auth_saml/controllers/main.py:
```diff
@@ -0,0 +1,216 @@
+                error = _("Sign up is not allowed on this database.")
+            elif error == '2':
+                error = _("Access Denied")
+            elif error == '3':
```
Why not use words to distiguish errors? It'll be easier to trace:

```suggestion
            if error == 'no-signup':
                error = _("Sign up is not allowed on this database.")
            elif error == 'access-denied':
                error = _("Access Denied")
            elif error == 'expired':
```

yajo commented on auth_saml/data/auth_saml.xml:
```diff
@@ -0,0 +1,84 @@
+034oo59YiLuo8FBVqYV1a3W0cWE9z1Qrn4CYiz9QQtbx04eKuEX+OxUCQQCzgGzi
+Oi41nsy8uFr5s4joYfdZfDzezLz/GZNJ1q950Z69nJRVVTbQCvyJ2/Q+MtvojDXl
+sIbngHtWnT36PTZS
+-----END PRIVATE KEY-----
```
What??? :scream: 

I hope this is demo data and you just have to label it as such...

yajo commented on auth_saml/doc/Makefile:
```diff
@@ -0,0 +1,238 @@
+# Makefile for Sphinx documentation
```
Please remove the whole `auth_saml/doc` folder.

Just extract the interesting stuff and put it in the README [after splitting it](https://github.com/OCA/maintainer-tools/tree/becfd07b7fdac38cb02a86a522c82d11b1c8e6fd/template/module/readme).

yajo commented on auth_saml/models/auth_saml.py:
```diff
@@ -0,0 +1,86 @@
+try:
+    import lasso
+except ImportError:
+    _logger.debug('Cannot `import lasso`.')
```
```suggestion
import lasso
```

No need to protect imports in v12+.

yajo commented on auth_saml/models/auth_saml.py:
```diff
@@ -0,0 +1,86 @@
+
+        self.ensure_one()
+
+        return self.matching_attribute
```
This method is not needed AFAICS. Just call `self.matching_attribute` and you're done.

yajo commented on auth_saml/models/auth_saml.py:
```diff
@@ -0,0 +1,86 @@
+    sp_pkey = fields.Text(
+        'Private key of our service provider (this openerpserver)'
+    )
+    matching_attribute = fields.Text(
```
If I'm not wrong, Text is too much here.

```suggestion
    matching_attribute = fields.Char(
```

yajo commented on auth_saml/models/auth_saml.py:
```diff
@@ -0,0 +1,86 @@
+
+    _name = 'auth.saml.provider'
+    _description = 'SAML2 provider'
+    _order = 'name'
```
```suggestion
    _order = 'sequence, name'
```

yajo commented on auth_saml/models/auth_saml.py:
```diff
@@ -0,0 +1,86 @@
+
+        login = self._get_lasso_for_provider()
+
+        # ! -- this is the part that MUST be performed on each call and
```
This is not XML!

```suggestion
        # This is the part that MUST be performed on each call and
```

yajo commented on auth_saml/models/auth_saml.py:
```diff
@@ -0,0 +1,86 @@
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+import logging
+import json as simplejson
```
Why? Just use it as `json` to avoid confusion:

```suggestion
import json
```

yajo commented on auth_saml/models/auth_saml.py:
```diff
@@ -0,0 +1,86 @@
+        login.initAuthnRequest()
+        login.request.nameIdPolicy.format = None
+        login.request.nameIdPolicy.allowCreate = True
+        login.msgRelayState = simplejson.dumps(state)
```
```suggestion
        login.msgRelayState = json.dumps(state)
```

yajo commented on auth_saml/models/auth_saml.py:
```diff
@@ -0,0 +1,86 @@
+        return login.msgUrl
+
+    # Name of the OAuth2 entity, authentic, xcg...
+    name = fields.Char('Provider name')
```
```suggestion
    name = fields.Char('Provider name', required=True)
```

yajo commented on auth_saml/models/auth_saml.py:
```diff
@@ -0,0 +1,86 @@
+
+    # Name of the OAuth2 entity, authentic, xcg...
+    name = fields.Char('Provider name')
+    idp_metadata = fields.Text('IDP Configuration')
```
What's IDP? Add help here.

yajo commented on auth_saml/models/auth_saml.py:
```diff
@@ -0,0 +1,86 @@
+    # Name of the OAuth2 entity, authentic, xcg...
+    name = fields.Char('Provider name')
+    idp_metadata = fields.Text('IDP Configuration')
+    sp_metadata = fields.Text('SP Configuration')
```
What's SP? Add help here.

yajo commented on auth_saml/models/auth_saml.py:
```diff
@@ -0,0 +1,86 @@
+    idp_metadata = fields.Text('IDP Configuration')
+    sp_metadata = fields.Text('SP Configuration')
+    sp_pkey = fields.Text(
+        'Private key of our service provider (this openerpserver)'
```
Shorten string (put i.e. `SP Private Key`), and then add a help with more details.

yajo commented on auth_saml/models/auth_saml.py:
```diff
@@ -0,0 +1,86 @@
+        default='subject.nameId',
+        required=True,
+    )
+    enabled = fields.Boolean('Enabled', default=False)
```
Rename this to let us use archivation:

```suggestion
    active = fields.Boolean(default=False)
```

yajo commented on auth_saml/models/auth_saml.py:
```diff
@@ -0,0 +1,86 @@
+    )
+    enabled = fields.Boolean('Enabled', default=False)
+    sequence = fields.Integer('Sequence')
+    css_class = fields.Char('CSS Class')
```
Add help here.

yajo commented on auth_saml/models/auth_saml.py:
```diff
@@ -0,0 +1,86 @@
+    enabled = fields.Boolean('Enabled', default=False)
+    sequence = fields.Integer('Sequence')
+    css_class = fields.Char('CSS Class')
+    body = fields.Char('Body')
```
Add help here.

yajo commented on auth_saml/data/ir_config_parameter.xml:
```diff
@@ -0,0 +1,11 @@
+
+    <record id="allow_saml_uid_and_internal_password" model="ir.config_parameter">
+        <field name="key">auth_saml.allow_saml.uid_and_internal_password</field>
+        <field name="value">0</field>
```
This file could be removed, as you can call `.get_param(name, 0)` and it will return `0` if missing.

yajo commented on auth_saml/models/base_settings.py:
```diff
@@ -0,0 +1,59 @@
+# Copyright (C) 2010-2016 XCG Consulting <http://odoo.consulting>
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+from odoo import api, fields, models
```
```suggestion
from odoo import api, fields, models
from odoo.tools import str2bool
```

yajo commented on auth_saml/models/base_settings.py:
```diff
@@ -0,0 +1,59 @@
+            return False
+
+        return (True if config_objs.value == '1' else False)
+
```
This is not needed:

```suggestion
```

yajo commented on auth_saml/models/base_settings.py:
```diff
@@ -0,0 +1,59 @@
+        res = super(ResConfigSettings, self).get_values()
+        get_param = self.env['ir.config_parameter'].sudo().get_param
+        res.update(allow_saml_uid_and_internal_password=get_param(
+            'auth_saml.allow_saml_uid_and_internal_password'))
```
```suggestion
        res.update(allow_saml_uid_and_internal_password=str2bool(get_param(
            'auth_saml.allow_saml_uid_and_internal_password', 'false'))
```

yajo commented on auth_saml/models/base_settings.py:
```diff
@@ -0,0 +1,59 @@
+            self.allow_saml_uid_and_internal_password = \
+                self.allow_saml_and_password()
+        set_param('auth_saml.allow_saml_uid_and_internal_password',
+                  repr(self.allow_saml_uid_and_internal_password))
```
```suggestion
                  self.allow_saml_uid_and_internal_password)
```

yajo commented on auth_saml/models/base_settings.py:
```diff
@@ -0,0 +1,59 @@
+
+_SAML_UID_AND_PASS_SETTING = 'auth_saml.allow_saml.uid_and_internal_password'
+
+
```
```suggestion
```

You're not using it now.

yajo commented on auth_saml/models/base_settings.py:
```diff
@@ -0,0 +1,59 @@
+    comments in the definition of the "res.config.settings" collection for
+    details.
+    """
+
```
This is quite obvious, you don't need it:

```suggestion
```

Remember about docs:

1. Make code easy to read, so you don't have to document *what* you're doing.
1. Instead, document *why* when it isn't obvious.

yajo commented on auth_saml/models/res_users.py:
```diff
@@ -0,0 +1,243 @@
+try:
+    import lasso
+except ImportError:
+    _logger.debug('Cannot `import lasso`.')
```
```suggestion
import lasso
```

No need to protect imports in v12+.

yajo commented on auth_saml/models/res_users.py:
```diff
@@ -0,0 +1,243 @@
+        if self._allow_saml_and_password():
+            pass
+
+        else:
```
```suggestion
        if not self._allow_saml_and_password():
```

yajo commented on auth_saml/models/res_users.py:
```diff
@@ -0,0 +1,243 @@
+                    self.id is not SUPERUSER_ID):
+                raise ValidationError(_("This database disallows users to "
+                                        "have both passwords and SAML IDs. "
+                                        "Errors for login %s" % (self.login)))
```
```suggestion
                                        "Errors for login %s") % self.login)
```

yajo commented on auth_saml/models/res_users.py:
```diff
@@ -0,0 +1,243 @@
+        # we are not yet logged in, so the userid cannot have access to the
+        # fields we need yet
+        login = p.sudo()._get_lasso_for_provider()
+        matching_attribute = p._get_matching_attr_for_provider()
```
```suggestion
        matching_attribute = p.matching_attribute
```

yajo commented on auth_saml/models/res_users.py:
```diff
@@ -0,0 +1,243 @@
+        try:
+            login.processAuthnResponseMsg(token)
+        except (lasso.DsError, lasso.ProfileCannotVerifySignatureError):
+            raise Exception('Lasso Profile cannot verify signature')
```
```suggestion
            raise AccessDenied('Lasso Profile cannot verify signature')
```

yajo commented on auth_saml/models/res_users.py:
```diff
@@ -0,0 +1,243 @@
+        except (lasso.DsError, lasso.ProfileCannotVerifySignatureError):
+            raise Exception('Lasso Profile cannot verify signature')
+        except lasso.ProfileStatusNotSuccessError:
+            raise Exception('Profile Status Not Success Error')
```
```suggestion
            raise AccessDenied('Profile Status Not Success Error')
```

yajo commented on auth_saml/models/res_users.py:
```diff
@@ -0,0 +1,243 @@
+        except lasso.ProfileStatusNotSuccessError:
+            raise Exception('Profile Status Not Success Error')
+        except lasso.Error as e:
+            raise Exception(repr(e))
```
Not needed:

```suggestion
```

yajo commented on auth_saml/models/res_users.py:
```diff
@@ -0,0 +1,243 @@
+                    name = attribute.name.decode('ascii')
+                except Exception as e:
+                    _logger.warning('sso_after_response: error decoding name \
+                        of attribute %s' % attribute.dump())
```
```suggestion
                        of attribute %s', attribute.dump())
```

yajo commented on auth_saml/models/res_users.py:
```diff
@@ -0,0 +1,243 @@
+                    except Exception as e:
+                        message = 'sso_after_response: name or format of an \
+                            attribute failed to decode as ascii: %s due to %s'
+                        _logger.warning(message % (attribute.dump(), str(e)))
```
```suggestion
                        _logger.warning(message, attribute.dump(), e)
```

yajo commented on auth_saml/models/res_users.py:
```diff
@@ -0,0 +1,243 @@
+                    except Exception as e:
+                        message = 'sso_after_response: value of an \
+                            attribute failed to decode as ascii: %s due to %s'
+                        _logger.warning(message % (attribute.dump(), str(e)))
```
```suggestion
                        _logger.warning(message, attribute.dump(), e)
```

yajo commented on auth_saml/models/res_users.py:
```diff
@@ -0,0 +1,243 @@
+        # production code...
+        assert len(user_ids) == 1
+        user = user_ids[0]
+
```
```suggestion
        user = self.search(
            [('saml_uid', '=', saml_uid), ('saml_provider_id', '=', provider)])
        if len(user) != 1:
            raise AccessDenied()
```

yajo commented on auth_saml/models/res_users.py:
```diff
@@ -0,0 +1,243 @@
+            if not res:
+                # TODO: maybe raise a defined exception instead of the last
+                # exception that occurred in our execution frame
+                raise
```
```suggestion
                raise AccessDenied()
```

yajo commented on auth_saml/models/res_users.py:
```diff
@@ -0,0 +1,243 @@
+                [('user_id', '=', self.env.user.id),
+                 ('saml_access_token', '=', token)])
+
+            # if the user is not found we re-raise the AccessDenied
```
```suggestion
```

yajo commented on auth_saml/models/res_users.py:
```diff
@@ -0,0 +1,243 @@
+        # - An SAML ID is being set.
+        # - The user is not the Odoo admin.
+        # - The "allow both" setting is disabled.
+        if (vals and vals.get('saml_uid') and self.id is not SUPERUSER_ID and
```
What if `saml_uid` was already present in the record?

You should use a constraint instead.

yajo commented on auth_saml/models/saml_token.py:
```diff
@@ -0,0 +1,30 @@
+        string="User",
+        # we want the token to be destroyed if the corresponding res.users
+        # is deleted
+        ondelete="cascade"
```
```suggestion
        ondelete="cascade",
        index=True,
```

yajo commented on auth_saml/models/res_users.py:
```diff
@@ -0,0 +1,243 @@
+        assert len(user_ids) == 1
+        user = user_ids[0]
+
+        # now find if a token for this user/provider already exists
+        token_ids = token_osv.search(
+            [('saml_provider_id', '=', provider), ('user_id', '=', user.id)])
+
+        if token_ids:
+            token_ids.write({'saml_access_token': saml_response})
+        else:
+            token_osv.create({'saml_access_token': saml_response,
+                              'saml_provider_id': provider,
+                              'user_id': user.id})
```
Is it possible for a user to have different tokens? For example if (s)he logs in in different devices...

If that's the case, then tokens should be always created, not written, and maybe they should have an expiration date, and a cron chould delete them after that.

yajo commented on auth_saml/scripts/2.0-cleanup.sql:
```diff
@@ -0,0 +1,2 @@
+--remove the old column from res.users
+ALTER TABLE res_users DROP COLUMN IF EXISTS saml_access_token;
```
Remove this.

yajo commented on auth_saml/scripts/clear_passwords.sql:
```diff
@@ -0,0 +1,5 @@
+-- ensure no Odoo user posesses both an SAML user ID and an Odoo password.
+-- Run this script to clear passwords of Odoo users that already have an SAML user ID.
+
+UPDATE res_users SET password = NULL WHERE password IS NOT NULL AND saml_uid IS NOT NULL;
```
:wastebasket:

yajo commented on auth_saml/security/ir.model.access.csv:
```diff
@@ -0,0 +1,4 @@
+id,name,model_id:id,group_id:id,perm_read,perm_write,perm_create,perm_unlink
+access_auth_saml_provider,auth_saml_provider,model_auth_saml_provider,base.group_system,1,1,1,1
+access_auth_saml_token,access_auth_saml_token,model_auth_saml_token,,0,0,0,0
```
Shouldn't you be able to read your own tokens? :thinking:

yajo commented on auth_saml/tests/test_auth_saml.py:
```diff
@@ -0,0 +1,23 @@
+        super(Test, self).setUp()
+        self.memory = TestMemory()
+
+    # TODO Tests.
```
Remove this file. :wastebasket: 

Or, better, add tests. :yum:

yajo commented on auth_saml/tests/util/odoo_tests.py:
```diff
@@ -0,0 +1,47 @@
+"""Utilities useful to Odoo tests.
```
:wastebasket:

yajo commented on auth_saml/tests/util/singleton.py:
```diff
@@ -0,0 +1,30 @@
+class Singleton(type):
```
:wastebasket:

yajo commented on auth_saml/tests/util/uuidgen.py:
```diff
@@ -0,0 +1,22 @@
+"""Utilities to handle unique ID generation.
```
:wastebasket:

yajo commented on auth_saml/views/auth_saml.xml:
```diff
@@ -0,0 +1,70 @@
+<?xml version="1.0"?>
+<odoo>
+
+    <!-- login form button, new in Odoo8, Odoo7 automatically adds the button -->
```
```suggestion
    <!-- login form button -->
```

yajo reviewed (changes requested):
Lots of things to fix/improve.

Also, rebase to fix conflicts please.

Thanks for your contribution!

yajo reviewed (changes requested):
Lots of things to fix/improve.

Also:

- Include newer history from 11.0 branch: https://github.com/OCA/server-auth/commits/11.0/auth_saml
- rebase to fix conflicts please.

Thanks for your contribution!

Commit:
[MIG][REF] General cleanup for migration

- Default behavior is now to allow password and SAML together.
  Otherwise, users could keep getting their passwords removed without warning.
- General cleanup.
- Remove relations to field `password_crypt` because in v12 the
  `password` field is always encrypted instead.

Co-Authored-By: Alexandre Díaz <alexandre.diaz@tecnativa.com>

yajo reviewed (approved):
I applied suggestions myself.

@Tecnativa TT22174

Tardo commented on auth_saml/controllers/main.py:
```diff
@@ -0,0 +1,210 @@
+
+        providers = self.list_providers()
+
+        response = super(SAMLLogin, self).web_login(*args, **kw)
```
```suggestion
        response = super().web_login(*args, **kw)
```

Tardo commented on auth_saml/controllers/main.py:
```diff
@@ -0,0 +1,210 @@
+            )
+
+        state = {
+            "d": request.session.db,
```
Can be good use other key names :/

yajo replied:
I think this is required by the SAML standard. It's being used as raw JSON somewhere.

Tardo commented on auth_saml/controllers/main.py:
```diff
@@ -0,0 +1,210 @@
+        """
+        saml_response = kw.get('SAMLResponse', None)
+
+        if kw.get('RelayState', None) is None:
```
By default is None
```suggestion
        if kw.get('RelayState') is None:
```

Tardo commented on auth_saml/controllers/main.py:
```diff
@@ -0,0 +1,210 @@
+        """client obtained a saml token and passed it back
+        to us... we need to validate it
+        """
+        saml_response = kw.get('SAMLResponse', None)
```
By default is None
```suggestion
        saml_response = kw.get('SAMLResponse')
```

Tardo commented on auth_saml/models/saml_token.py:
```diff
@@ -0,0 +1,27 @@
+        ondelete="cascade",
+        index=True,
+    )
+    saml_access_token = fields.Char(
```
This should be required no?

yajo replied:
I guess all 3 fields should. :thinking:

Tardo commented on auth_saml/security/auth_saml.xml:
```diff
@@ -0,0 +1,84 @@
+</EntityDescriptor>]]>
+        </field>
+        <field name="sp_pkey"><![CDATA[
+-----BEGIN PRIVATE KEY-----
```
I don't know what this cert is used for.... but this doesn't looks very private :P

yajo replied:
:rofl: Lol true, I forgot to move this to demo data.

Tardo reviewed (changes requested):
Looks good! only few comments....

yajo reviewed (approved):
All attended.

yajo reviewed (approved):
This is now ready to merge.

gurneyalex reviewed (commented):
approved --but can't self approve -- 

code review, no test

pedrobaeza commented:
@Yajo, please check Travis

yajo commented:
All green now.

pedrobaeza commented:
/ocabot merge

NuevaIT-ext1 commented:
i am new here and i found and error in auth_saml.py,line 50 lasso.Server
module 'lasso' has no attribute 'Server'

eLBati commented:
@NuevaIT-ext1 you need to install the "Free Liberty Alliance Implementation" https://lasso.entrouvert.org/

https://packages.ubuntu.com/focal/python3-lasso
